coderd/database/queries/templateversions.sql

```diff
@@ -129,3 +136,91 @@ WHERE
 	AND template_id = $3
 ORDER BY created_at DESC
 LIMIT 1;
+
+-- name: UnarchiveTemplateVersion :exec
+-- This will always work regardless of the current state of the template version.
+UPDATE
+	template_versions
+SET
+	archived = false,
+	updated_at = sqlc.arg('updated_at')
+WHERE
+		id = sqlc.arg('template_version_id');
+
+-- name: ArchiveUnusedTemplateVersions :many
```

Comment:
Would it be simpler to do this in a tx version-by-version at the Go level instead?

My reasoning is that there's likely to be an initial deluge of archived template versions when this is first performed.

We can more easily rate-limit this at the Go layer -- if we're archiving in the background, we can archive up to X, wait a while, archive up to another X, and so on until there are no more to archive.

Reply:
The downside of doing it in Go is you have to handle the possible state changes going on in the background. I put it all in postgres to avoid these issues.

As for how costly it will be, I can't imagine this query being that costly that we need to break it up into multiple queries. Like a few seconds isn't that bad. We do these kinds of joins when we fetch "all workspaces" to see their status, so this kind of joining isn't unprecedented.

Reply:
Fair. If we need to break it up later, we can do so.